From 78ab736042c941844b067a0e05b819fdd036eba9 Mon Sep 17 00:00:00 2001
From: Ciprian Marian Costea <ciprianmarian.costea@nxp.com>
Date: Fri, 11 Jun 2021 16:07:33 +0300
Subject: [PATCH 12/14] s32g274a: Add driver as part of LPDDR4 ERR050543

This simple driver will export the value of 'polling_needed'
variable in order to be used by the Linux driver through
later dtb fixup.

Upstream-Status: Pending

Issue: ALB-5558
Signed-off-by: Ciprian Marian Costea <ciprianmarian.costea@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 arch/arm/cpu/armv8/s32/Makefile         |  1 +
 arch/arm/cpu/armv8/s32/fdt.c            |  7 +++--
 arch/arm/dts/fsl-s32g.dtsi              |  6 +++++
 board/freescale/Kconfig                 |  1 +
 drivers/ddr/fsl/Makefile                |  5 ++++
 drivers/ddr/fsl/ddr_s32gen1_err050543.c | 36 +++++++++++++++++++++++++
 drivers/ddr/fsl/ddr_s32gen1_err050543.h | 11 ++++++++
 7 files changed, 63 insertions(+), 4 deletions(-)
 create mode 100644 drivers/ddr/fsl/ddr_s32gen1_err050543.c
 create mode 100644 drivers/ddr/fsl/ddr_s32gen1_err050543.h

diff --git a/arch/arm/cpu/armv8/s32/Makefile b/arch/arm/cpu/armv8/s32/Makefile
index dcc1bf415b..ff2a3d2c78 100644
--- a/arch/arm/cpu/armv8/s32/Makefile
+++ b/arch/arm/cpu/armv8/s32/Makefile
@@ -23,3 +23,4 @@ obj-$(CONFIG_FSL_DCU_FB)	+= dcu.o
 obj-y				+= initsram.o
 ccflags-y			+= -Idrivers/clk/s32/include
 ccflags-y			+= -Iboard/freescale/s32-gen1
+ccflags-$(CONFIG_SYS_ERRATUM_ERR050543)	+= -Idrivers/ddr/fsl
diff --git a/arch/arm/cpu/armv8/s32/fdt.c b/arch/arm/cpu/armv8/s32/fdt.c
index 9e858ec07a..4488f4c0c3 100644
--- a/arch/arm/cpu/armv8/s32/fdt.c
+++ b/arch/arm/cpu/armv8/s32/fdt.c
@@ -17,6 +17,9 @@
 #include <dt-bindings/phy/phy.h>
 #include <linux/ctype.h>
 #include "mp.h"
+#ifdef CONFIG_SYS_ERRATUM_ERR050543
+#include <ddr_s32gen1_err050543.h>
+#endif
 
 #define ID_TO_CORE(ID)	(((ID) & 3) | ((ID) >> 7))
 
@@ -43,10 +46,6 @@
 
 #define S32_DDR_LIMIT_VAR "ddr_limitX"
 
-#ifdef CONFIG_SYS_ERRATUM_ERR050543
-extern uint8_t polling_needed;
-#endif
-
 #ifdef CONFIG_MP
 
 #if CONFIG_S32_ATF_BOOT_FLOW
diff --git a/arch/arm/dts/fsl-s32g.dtsi b/arch/arm/dts/fsl-s32g.dtsi
index f98616d910..ef72ee8325 100644
--- a/arch/arm/dts/fsl-s32g.dtsi
+++ b/arch/arm/dts/fsl-s32g.dtsi
@@ -56,6 +56,12 @@
 		status = "okay";
 	};
 
+	ddr: ddr@403C0000 {
+		compatible = "fsl,s32gen1-ddr-err050543";
+		/* Will be modified by ATF */
+		status = "disabled";
+	};
+
 	pcie1: pcie@44100000 {
 		compatible = "fsl,s32gen1-pcie";
 		reg =
diff --git a/board/freescale/Kconfig b/board/freescale/Kconfig
index e84ed78b65..cb8dccb2f2 100644
--- a/board/freescale/Kconfig
+++ b/board/freescale/Kconfig
@@ -5,6 +5,7 @@ config S32V2
 config S32_GEN1
 	bool "S32 GEN1 Common Chassis"
 	imply CMD_GPIO
+	imply MISC
 	select ARM_SMCCC
 	select CLK
 	select CLK_SCMI
diff --git a/drivers/ddr/fsl/Makefile b/drivers/ddr/fsl/Makefile
index e28ef6cb68..ad326122f5 100644
--- a/drivers/ddr/fsl/Makefile
+++ b/drivers/ddr/fsl/Makefile
@@ -1,6 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 #
 # Copyright 2008-2014 Freescale Semiconductor, Inc.
+# Copyright 2021 NXP
 
 obj-$(CONFIG_SYS_FSL_DDR1) += main.o util.o ctrl_regs.o options.o \
 				lc_common_dimm_params.o
@@ -38,3 +39,7 @@ obj-$(CONFIG_SYS_FSL_DDRSS) += ddrss_emu.o
 else
 obj-$(CONFIG_SYS_FSL_DDRSS) += ddrss.o
 endif
+
+ifdef CONFIG_S32_ATF_BOOT_FLOW
+obj-$(CONFIG_SYS_ERRATUM_ERR050543) += ddr_s32gen1_err050543.o
+endif
diff --git a/drivers/ddr/fsl/ddr_s32gen1_err050543.c b/drivers/ddr/fsl/ddr_s32gen1_err050543.c
new file mode 100644
index 0000000000..f8fa1c5f34
--- /dev/null
+++ b/drivers/ddr/fsl/ddr_s32gen1_err050543.c
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright 2021 NXP
+ *
+ * This driver serves in context of NXP SYS_ERRATUM_ERR050543.
+ * It is used only in ATF boot flow.
+ * During binding, it sets and exports the value of 'polling_needed'
+ * variable in order to apply Linux dtb fixup logic signaling
+ * that ERR050543 workaround has been applied in ATF and needs
+ * to be reverted, when possible, by the associated Linux driver.
+ */
+#include <common.h>
+#include <dm.h>
+#include "ddr_s32gen1_err050543.h"
+
+static int s32gen1_ddr_bind(struct udevice *dev)
+{
+	/* Hook in order to set the 'polling_needed' variable value
+	 * in case of ATF boot flow.
+	 */
+	polling_needed = 1;
+
+	return 0;
+}
+
+static const struct udevice_id s32gen1_ddr_ids[] = {
+	{ .compatible = "fsl,s32gen1-ddr-err050543" },
+	{ /* sentinel */ }
+};
+
+U_BOOT_DRIVER(ddr_s32gen1) = {
+	.name = "s32gen1_ddr",
+	.id = UCLASS_MISC,
+	.of_match = s32gen1_ddr_ids,
+	.bind = s32gen1_ddr_bind,
+};
diff --git a/drivers/ddr/fsl/ddr_s32gen1_err050543.h b/drivers/ddr/fsl/ddr_s32gen1_err050543.h
new file mode 100644
index 0000000000..e5c9da3242
--- /dev/null
+++ b/drivers/ddr/fsl/ddr_s32gen1_err050543.h
@@ -0,0 +1,11 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Copyright 2021 NXP
+ */
+
+#ifndef __S32GEN1_DDR_H_
+#define __S32GEN1_DDR_H_
+
+extern u8 polling_needed;
+
+#endif /* __S32GEN1_DDR_H_ */
-- 
2.17.1

